test/hotspot/jtreg/serviceability/jvmti/vthread/PopFrameTest/PopFrameTest.java

```diff
@@ -0,0 +1,252 @@
+ *     Its method run() invokes the following methods:
+ *      - method A() that is blocked on a monitor
+ *      - method B() that is stopped at a breakpoint
+ *      - method C() that forces agent to call PopFrame on its own thread
+ *     JVMTI PopFrame is called in all cases.
+ */
+public class PopFrameTest {
+    private static final String agentLib = "PopFrameTest";
+    static final int JVMTI_ERROR_NONE = 0;
+    static final int THREAD_NOT_SUSPENDED = 13;
+    static final int OPAQUE_FRAME = 32;
+    static final int PASSED = 0;
+    static final int FAILED = 2;
+
+    static void log(String str) { System.out.println(str); }
+
+    static native void prepareAgent(Class taskClass);
+    static native void suspendThread(Thread thread);
+    static native void resumeThread(Thread thread);
+    static native void ensureAtBreakpoint();
+    static native void notifyAtBreakpoint();
+    static native int  popFrame(Thread thread);
+
+    static int status = PASSED;
+    static boolean is_virtual = true;
+
+    static void setFailed(String msg) {
+        log("\nFAILED: " + msg);
+        status = FAILED;
+    }
+
+    static void throwFailed(String msg) {
+        log("\nFAILED: " + msg);
+        throw new RuntimeException("PopFrameTest failed!");
+    }
+
+    public static void main(String args[]) {
+        is_virtual = !(args.length > 0 && args[0].equals("platform"));
+        run();
+        if (status == FAILED) {
+            throwFailed("PopFrameTest!");
+        }
+        log("\nPopFrameTest passed");
+    }
+
+    public static void run() {
+        TestTask testTask = new TestTask();
+        Thread testTaskThread = null;
+        int errCode;
+
+        prepareAgent(TestTask.class);
+
+        log("\nMain #A: method A() must be blocked on entering a synchronized statement");
+        if (is_virtual) {
+            testTaskThread = Thread.ofVirtual().name("TestTaskThread").start(testTask);
+        } else {
+            testTaskThread = Thread.ofPlatform().name("TestTaskThread").start(testTask);
+        }
+        testTask.ensureStarted();
+
+        log("\nMain #A.1: unsuspended");
+        {
+            errCode = popFrame(testTaskThread);
+            if (errCode != THREAD_NOT_SUSPENDED) {
+                throwFailed("Main #A.1: expected THREAD_NOT_SUSPENDED instead of: " + errCode);
+            } else {
+                log("Main #A.1: got expected THREAD_NOT_SUSPENDED");
+            }
+
+            log("\nMain #A.2: suspended");
+            suspendThread(testTaskThread);
+            errCode = popFrame(testTaskThread);
+            if (errCode != JVMTI_ERROR_NONE) {
+                throwFailed("Main #A.2: expected JVMTI_ERROR_NONE instead of: " + errCode);
+            } else {
+                log("Main #A.2: got expected JVMTI_ERROR_NONE");
+            }
+            resumeThread(testTaskThread);
+            testTask.clearDoLoop();
+            testTask.sleep(5);
```

Comment:
Why sleep is needed here?

Reply:
It is to better sync the output between the main and target threads.
It becomes better ordered and understandable.